### ISSUE TITLE
ZEC <-> XLM Service

### ISSUE THREAD

**Issue by zallarak** [ZcashFoundation/GrantProposals-2017Q4#18]
# ZEC <-> XLM Service

## Motivation and overview

The overarching motivation is widespread adoption of cryptocurrencies and privacy-focused cryptocurrencies, to protect our civil liberties.

Stellar's built-in exchange and Anchor functionality has made it an ideal choice for remittance and money transfer in general.

Zcash's privacy and selective transparency of transactions makes it an attractive privacy-enabled currency to integrate into the Stellar network. The ZEC <-> XLM service would publicly auditable to its users, and subsequently, users may transfer money out of internal wallets via private transactions, providing them the privacy they wanted.

## Technical approach

- Web API: exposing routes to authenticate and initiate wallet transactions
- Asynchronous job queue to monitor users' accounts for incoming / outgoing transaction statuses
- Data store: Secure storage of users' Zcash and Stellar private keys
- Zcash + Stellar nodes
- Front end user interface

## Evaluation plan

Fully functional Zcash and Stellar web-based wallet (much like Coinbase's Wallet).

- Become a Zcash Anchor on the Stellar Network
- 1000 users within 3 months of launching all features
- Enable anonymous remittance
- $1,000,000 USD of monthly transactions within 1 year

## Security considerations

- **External Security Audits and penetration testing**
- Web-based wallets carry an inherent risk; you are trusting them with your secret key(s)
- 2 factor auth and strong password requirements
- Zcash and Stellar reserve funds should be held in offline wallets, and transactions should be programatically created and initiated in a way such that private keys for large wallets are not exposed to the internet
- As a web-wallet for cryptocurrencies, we would be subject to attacks

## Schedule

1. Launch Stellar Wallet functionality (0.5 months)
2. Launch Zcash Wallet functionality for public transactions (1 month)
3. Support private Zcash transactions (1 month)
4. Become an official Zcash Stellar Anchor, and allow for seamless ZEC <-> XLM exchange (1 month)
5. Security auditing and penetration testing (throughout)

## Budget and justification (approx.)

$13,000 USD

- Time compensation: $5,000 USD
- Design contracting fees: $5,000 USD
- Security audit / penetration testing: $2,000
- Hosting / External Fees Fund: $1,000 USD

**Comment by arcalinea** [comment]
Is there an upper bound for grant proposals? Large budgets for funding entirely new projects like this do not seem justified at this time, when there are so many improvements that can be made by members of the community who have already been contributing. 

Besides, just as a general principle, a Foundation should not be funding a full-time developer to live in the most expensive city in the US for a few months to work on an experimental project, this sounds more like something that wants seed funding.

**Comment by zallarak** [comment]
@arcalinea thanks for responding. I misunderstood the budget and thought $80,000 USD was set aside for each grant proposal, when in reality that is the total fund.

I agree that it was out of line and have revised it. I think I could make ends meet by consulting in parallel.

Does it seem more reasonable now? Thanks again for pointing this out. cc @tromer

**Comment by zallarak** [comment]
We've made a lot of progress on the Stellar Wallet functionality, here are some rough UI mockups. After shipping that, we'd concentrate on supporting Zcash. Having a way to move from Zcash to Stellar seamlessly opens the door for exchanging with tons of other currencies on the Stellar network.

![dashboard](https://user-images.githubusercontent.com/806155/30817982-2d48344c-a1cf-11e7-854f-4cf7ae053fba.jpg)

**Comment by acityinohio** [comment]
Every informal proposal has multiple reviews by the review committee. The reviews are being collected and discussed in a private google doc (the 5 reviewers all have edit access to it, no one else can view it). By way of early, informal feedback, the reviewers would like to inform **you that your proposal is not a leading candidate for funding in this round.**

You may still choose to submit a full proposal by **October 6th**, and it will be evaluated. But if it’s substantially the same as discussion so far reflects, then we suspect it is unlikely to be chosen for funding; and if it isn't, then we encourage you to post a draft as early as possible, to allow for community feedback. Regardless of your choice we thank you for participation thus far.

**Comment by zallarak** [comment]
Thanks for the notification and feedback.

**Comment by zookozcash** [comment]
@zallarak: for your information, I loved this proposal, and I hope you, or someone, can do something like this someday. :-)

(I'm not part of the Zcash Foundation and I'm not part of the board judging these proposals.)

**Comment by zallarak** [comment]
@acityinohio why did you re-open this?

Also, I ended up shipping the [web Stellar wallet](https://astralwallet.io) and we have users. We're shortly adding support for stellar-based tokens.

In order to make the ZEC <-> XLM service, the next steps would presumably be:

- become a ZEC anchor and issue ZEC tokens
- create a way to redeem ZEC from ZEC tokens (I believe Stellar already has a spec for how to do this, we'd just need to implement it)

Here's a screenshot of our next iteration:
![astral tokens](https://user-images.githubusercontent.com/806155/33087566-d3f6d742-ce9f-11e7-99a5-48f3801df933.png)
